### ISSUE TITLE
tables.pl module: explicit left align doesn't add align="left", alignment markup leaves spurious whitespace in HTML

### ISSUE THREAD

**Issue by greenfive** [kensanata/oddmuse#13]
**Description:**

According to https://oddmuse.org/wiki/Table_Markup_Extension one can explicitly align the content of table cells with

- leading white space = align right
- trailing white space = align left
- leading+trailing white space = align center

This doesn't work correctly for leading white space, which doesn't add align="left" to `<td>` or `<th>`.

**How to repeat:**

```
|=left |= middle |= right|
|left | middle | right|
|xxxxxxxxxxxxxxxxxxxxxx|xxxxxxxxxxxxxxxxxxxxxx|xxxxxxxxxxxxxxxxxxxxxx|
```

which yields

`<table class="user"><tr><th>left </th><th align="center">middle </th><th align="right">right</th></tr><tr><td>left </td><td align="center">middle </td><td align="right">right</td></tr><tr><td>xxxxxxxxxxxxxxxxxxxxxx</td><td>xxxxxxxxxxxxxxxxxxxxxx</td><td>xxxxxxxxxxxxxxxxxxxxxx</td></tr></table>`

1. `<th>left </th>` misses explicit `align="left"` and has additional trailing white space not chopped
2. `<td>left </td>` misses explicit `align="left"` and has additional trailing white space not chopped
3. `<th align="center">middle </th>` has additional trailing white space not chopped
4. `<td align="center">middle </td>` has additional trailing white space not chopped

**Expected:**

`<table class="user"><tr><th align="left">left</th><th align="center">middle</th><th align="right">right</th></tr><tr><td align="left">left</td><td align="center">middle</td><td align="right">right</td></tr><tr><td>xxxxxxxxxxxxxxxxxxxxxx</td><td>xxxxxxxxxxxxxxxxxxxxxx</td><td>xxxxxxxxxxxxxxxxxxxxxx</td></tr></table>`

**Additional notes:**

I think the problem is in `sub TableAttributes` in `tables.pl`.

**Comment by kensanata** [comment]
I think the problem is that your tables are being generated by a different module. `tables.pl` uses two vertical bars, like the old UseMod markup (`||`). I looked at the tests in `tables.t` and they are testing for left, right and center alignment, and they pass. I'm guessing that this unknown other module you are using is in fact broken, but which is it? `creole.pl`? I just looked at `creole.t` and noticed that it contains `align="center"` and `align="right"` but no `align="left"`. So that's a candidate.

As for the trailing single space, I simply decided to accept it, so usually the tests already contain it. But I'm not opposed to somebody fixing it. :)

**Comment by greenfive** [comment]
You are perfectly right. In fact I even didn't use `tables.pl` when I found that bug, as it is stated on the modules documentation page that one should not use it together with `creole.pl`, which is the main markup module I use.While I tried to figure out the cause of that bug, I used `table.pl` as an additional temporary drop in, but than my markup with double pipes "||" for tables got (as I assume) parsed by the creole module, which didn't give any usable results.

I had a thorough look at `creole.pl` and found the following lines:

```
         if ($is_left_justified and
             $is_right_justified) { $attributes .= 'align="center"' }
      elsif ($is_right_justified) { $attributes .= 'align="right"' }
      # this is the default:
      # elsif ($is_left_justified) { $attributes .= 'align="left"' }

```

It looked suspicious to me that the last elsif branch is commented, exactly where the test for just left justification is made and `align="left"` would get added. For a test I uncommented that last elsif branch in my local installation and it started to work – at least as far as I tested – correctly, `align="left"` got added to both `th` and `td` cells where it is needed.
However there must be a reason why that last branch was commented. I don't want to submit a patch, before someone with deeper markup/creole knowledge checks for unwanted side effects or knows the reason why it was commented first hand.

**Comment by kensanata** [comment]
I guess people felt that cells are left aligned by default in every browser
and thus no extra CSS attribute was needed. Except that header cells are
centered by default, I think.

**Comment by greenfive** [comment]
Yes, `th` are rendered centered by default and that was the root cause of my investigation, as I tried to change it. Do you mind making the change to `creole.pl` or should it be configurable?
I'd vote for the change, not making it configurable though, because if one adds a space character to have it explicitly aligned, it's on purpose.

**Comment by kensanata** [comment]
I think this should be fixed.

**Comment by kensanata** [comment]
Fixed in 60a4572, I hope?

**Comment by greenfive** [comment]
Just for the record: Applied the patch to all of my wikis and checked it thoroughly. The patch has solved the issue in all cases I formerly had problems. No side effects noted.
Thanks alot for your effort, Alex!
